### PR TITLE
Test on macOS-13 (x86) and macOS-14 (arm)

### PR DESCRIPTION
**Pull request by simlay**
[Github now has m1 runners](https://github.blog/changelog/2024-01-30-github-actions-introducing-the-new-m1-macos-runner-available-to-open-source/). Using macOS-14 will exclusively run on M1s while macOS-13 will still be on intel.

### CODE REVIEW

**Comment by simlay** [comment]
@kali I see that you retried the CI run. I authored this PR hoping that the issues I was having in my CI's setup were specific to my project. I authored [a repo and CI hoping it was the runner itself](https://github.com/simlay/macos-14-runner-simctl-launch-issue/blob/main/.github/workflows/macos.yml) to file against [the runner-images repo](https://github.com/actions/runner-images/).

I'm not sure where to go from here.

**Comment by briansmith** [comment]
It would make more sense to me to target aarch64-apple-ios when the host is aarch64-apple-darwin, and x86_64-apple-ios when the host is x86_64-apple-darwin.

File: .travis.sh [hunk 1]
```diff
@@ -67,12 +67,12 @@ then
      title "boot a simulator"
      rustup target add x86_64-apple-ios;
      RUNTIME_ID=$(xcrun simctl list runtimes | grep iOS | cut -d ' ' -f 7 | tail -1)
-     export SIM_ID=$(xcrun simctl create My-iphone7 com.apple.CoreSimulator.SimDeviceType.iPhone-8 $RUNTIME_ID)
+     export SIM_ID=$(xcrun simctl create My-iphone-se com.apple.CoreSimulator.SimDeviceType.iPhone-SE-3rd-generation $RUNTIME_ID)
```
Comment (briansmith) [inline .travis.sh:99]: Please add a comment on why we are using this specific choice of iPhone to simulate. If iPhone 8 works when running on M1 then maybe this is better done in a separate PR?

**Comment by simlay** [comment]
> It would make more sense to me to target aarch64-apple-ios when the host is aarch64-apple-darwin, and x86_64-apple-ios when the host is x86_64-apple-darwin.

Turns out you're right. Maybe rosetta-2 isn't in the m1 runners or has some quirks.

File: .travis.sh [hunk 2]
```diff
@@ -60,19 +60,52 @@ tests_sequence() {
...
      title "••••• Darwin: ios simulator tests •••••"
      title "boot a simulator"
-     rustup target add x86_64-apple-ios;
+     rustup target add x86_64-apple-ios aarch64-apple-ios-sim;
      RUNTIME_ID=$(xcrun simctl list runtimes | grep iOS | cut -d ' ' -f 7 | tail -1)
-     export SIM_ID=$(xcrun simctl create My-iphone7 com.apple.CoreSimulator.SimDeviceType.iPhone-8 $RUNTIME_ID)
+     export SIM_ID=$(xcrun simctl create My-iphone-se com.apple.CoreSimulator.SimDeviceType.iPhone-SE-3rd-generation $RUNTIME_ID)
      xcrun simctl boot $SIM_ID
-     tests_sequence $SIM_ID
+     if [ "$(uname -m)" = "arm64" ]; then
```
Comment (briansmith) [inline .travis.sh:101]: I suggest you put `rustup target add aarch64-apple-ios-sim` here and remove it above. This way we only install the relevant target.

File: .travis.sh [hunk 3]
```diff
@@ -60,19 +60,52 @@ tests_sequence() {
...
+         tests_sequence_ios_sim $SIM_ID
+     else
```
Comment (briansmith) [inline .travis.sh:103]: Ditto: Put `rustup target add x86_64-apple-ios` here and remove it above.

File: .travis.sh (hunk 2 above)
Comment (briansmith) [inline .travis.sh:101]: I suggest you add a comment above this, something like:
```
# The x86_64-apple-ios target seems to not work on an ARM64 host,
# and the aarch64-apple-ios-sim target doesn't work on an x86-64 host.

File: .travis.sh [hunk 4]
```diff
@@ -60,19 +60,52 @@ tests_sequence() {
     )
 }
 
+tests_sequence_ios_sim() {
```
Comment (briansmith) [inline .travis.sh:63]: It seems like this should have `aarch64` or `arm64` in the name, and `test_sequence` should have `x86_64` in the name.

File: .travis.sh [hunk 5]
```diff
@@ -60,19 +60,52 @@ tests_sequence() {
...
+    title "testing from workspace directory"
+    ( \
+        cd test-ws \
+        && cargo clean \
+        && $CARGO_DINGHY   -d $1 -p auto-ios-aarch64-sim test pass \
+        && ! $CARGO_DINGHY -d $1 -p auto-ios-aarch64-sim test fails \
+        && ! $CARGO_DINGHY -d $1 -p auto-ios-aarch64-sim test \
```
Comment (briansmith) [inline .travis.sh:70]: I am new to this tool so I don't know what `-p` does or why it seems to be needed here. I think maybe we're working around a bug where cargo-dinghy isn't automatically choosing `-sim` when the host is aarch64, even though it does automatically choose the right thing for x86_64?

if this is indeed a bug, perhaps we should fix that bug so that we can DRY these two functions?

**Comment by briansmith** [comment]
> Turns out you're right. Maybe rosetta-2 isn't in the m1 runners or has some quirks.

The symptom is an indefinite hang, right? Maybe macOS is showing the GUI prompt about installing Rosetta, and that's what's causing the hang?

File: .travis.sh (hunk 5 above)
Comment (simlay) [reply .travis.sh:70]: I was just wondering about that. It's a bug. To verify locally, I ran `rustup target remove x86_64-appl-ios` (on my m1), then `cargo dinghy -d <sim id> test` and it failed because dinghy was compiling for `x86_64-apple-ios` rather than `aarch64-apple-ios-sim`

File: .travis.sh (hunk 4 above)
Comment (simlay) [reply .travis.sh:63]: Yeah, I don't love it. The `test_sequence` function is used for android, qemu and the iOS targets. I think `tests_sequence_aarch64_ios_sim` might be okay but it's pretty un-DRY.

File: .travis.sh (hunk 5 above)
Comment (briansmith) [reply .travis.sh:70]: > cargo dinghy -d <sim id> test and it failed because dinghy was compiling for x86_64-apple-ios rather than aarch64-apple-ios-sim

Maybe the best thing to do by default is query the target device and choose the architecture that matches the host, if available, and then choose the default target triple based on what's compatible for that device? If the target device is a real device then it would be aarch64-apple-ios and if it is a simulator then either aarch64-apple-ios-sim or x86_64-apple-ios? And do analogously if the target device is a tvos (simulator) or watchos (simulator).

File: .travis.sh (hunk 4 above)
Comment (briansmith) [reply .travis.sh:63]: Ah, I see. I agree that's a better name.

File: .travis.sh (hunk 5 above)
Comment (briansmith) [reply .travis.sh:70]: Also, I think the kind of improvement I'm suggesting in this thread could be postponed to a future PR. The duplication here is unfortunate but this is good progress that would unblock work on several issues.

**Review by briansmith — APPROVED** [review]
I have no authority in this project, but LGTM. Thanks for doing this!

**Comment by simlay** [comment]
> I have no authority in this project, but LGTM. Thanks for doing this!

Thanks for the review! I'm glad to hear getrandom (and perhaps ring) is using dinghy! 

@kali could you review/merge this when you get a chance?

**Comment by briansmith** [comment]
> The symptom is an indefinite hang, right? Maybe macOS is showing the GUI prompt about installing Rosetta, and that's what's causing the hang?

That can't be the problem because I've run the prebuild x86_64 binary on a macos-14 runner.

**Review by kali — APPROVED** [review]

**Comment by kali** [comment]
Thanks!

The -p/--platform is needed because, `cargo dinghy` will by default build for the host. This is by design, but it may not have been the best idea.
